File: cxx/traaxx/sources/propagate_up_single_pass.cpp
#include <traaxx/propagate_up_single_pass.hpp>

namespace traaxx
{
    template<typename IndexT>
    BitMask propagate_up_single_pass(const std::vector<IndexT> &parent, BitMask seed)
    {
        auto const n = static_cast<IndexT>(parent.size());
        for (IndexT i = n; i-- > 0;)
        {
            if (seed.get(i))
            {
                seed.set(parent[i]);
            }
        }
        return seed;
    }

    template BitMask propagate_up_single_pass<std::uint32_t>(const std::vector<std::uint32_t> &, BitMask);
}

File: go/std/propagate_up_single_pass.go
package std

// PropagateUpSinglePass propagates a mask upward in a single sequential
// pass, in reverse topological order. Mirrors
// traaxx::propagate_up_single_pass
// (cxx/traaxx/sources/propagate_up_single_pass.cpp). Clones seed before
// mutating it (see PropagateDownSinglePass for why).
func PropagateUpSinglePass(parent []Index, seed BitMask) BitMask {
	result := seed.Clone()
	n := len(parent)
	for i := n; i > 0; i-- {
		idx := i - 1
		if result.Get(Index(idx)) {
			result.Set(parent[idx])
		}
	}
	return result
}
